### ISSUE TITLE
How to interpret CrcFile values

### ISSUE THREAD

**Issue by ThomasBreuer** [gap-system/gap#4164]
I have the values of GAP's `CrcFile` function for several files.
Now I want to use these files in a context independent of GAP, for example in Julia, and to use also the known checksums.
Julia's `CRC32c` package provides a function `crc32c` for computing a CRC value of a file, but the results are different from those of GAP's `CrcFile`.
Julia's `crc32c` function admits an optional argument, can one perhaps use it to get results compatible with those of GAP's `CrcFile`, or is it necessary to reimplement GAP's `CrcFile` in Julia?
(I think it would be useful to document what exactly GAP's `CrcFile` does.)

**Comment by fingolfin** [comment]
There are multiple caveats with CRC32: First of, the initial seed value. GAP seems to use 0x12345678; Wikipedia suggests 0xFFFFFFFF; Julia might be using 0 (if I read its documentation and code right). Just trying `0x12345678` as second argument for `crc32c` doesn't do the trick, though.

The next pitfall is bigendian vs little endian. And also that there are other minor variants of CRC32 that float around.

As a further complication, GAP seems to treat newline in a special way (in an attempt to treat Windows and Unix line ends the same -- ugh). 

I played a bit around, but couldn't find a seed that worked in any meaningful way (i.e. even after applying LE <-> BE conversions to the CRC32 values). Here are some naive things I tried: First, get some reference values from GAP:
```gap
gap> List(["", "a", "A", "aA"], x -> HexStringInt(CrcString(x)));
[ "1234567", "64799C9", "5F17BC5", "18D33EE" ]
```
versus:
```julia
julia> using CRC32c

julia> for i = 1:2^32
         seed = UInt32(i)
         if crc32c("", seed) == 0x01234567
            println(seed, " is the right seed")
            break
         end
       end
19088743 is the right seed

julia> seed
0x01234567

julia> [crc32c(x, seed) for x in ["", "a", "A", "aA"]]
4-element Array{UInt32,1}:
 0x01234567
 0x74dd97fc
 0x54601922
 0x5f34b0dc

julia> for i = 1:2^32
         seed = UInt32(i)
         if crc32c("", seed) == 0x67452301
            println(seed, " is the right seed")
            break
         end
       end
1732584193 is the right seed

julia> seed
0x67452301

julia> [crc32c(x, seed) for x in ["", "a", "A", "aA"]]
4-element Array{UInt32,1}:
 0x67452301
 0x33dc8510
 0x13610bce
 0xf100d604

```

Depending on your usecase, it might be better to use a modern and standardized hash like SHA2-256 (or if you speed is critical and security not, then e.g. MD5).

**Comment by ThomasBreuer** [comment]
@fingolfin Thanks.
My current usecase is the Atlas of Group Representations (plus extensions). Some time ago, I had added GAP's `CrcFile` values to the files that list the contents of this database, in order to check whether a downloaded file looks reasonable.
I understand your last sentence in such a way that I should better forget about `CrcFile`, and use a real standard.

**Comment by fingolfin** [comment]
I actually forgot to mention the biggest source of differences: there are many different polynomials one can use for CRC32, there is no single standard. Julia uses 0x82f63b78 aka CRC-32C (Castagnoli) (cf. the table on https://en.wikipedia.org/wiki/Cyclic_redundancy_check). They are doing this because modern Intel CPUs have a CPU instruction (as part of SSE4.2) supporting this.

The GAP implementation doesn't explicitly say which polynomial it uses, but my guess would be that it uses "plain" CRC-32 as used in Z-MODEM and other old and new standards (I tried using https://crccalc.com to figure out which it is, but the fact that GAP uses a nonstandard initial value makes it difficult to do this way. Anyway, doesn't matter)

So In a nutshell: yes, forget about it.

**Comment by ThomasBreuer** [comment]
Thanks a lot.
